core/include/assert.h

```diff
@@ -41,10 +65,19 @@ extern const char assert_crash_message[];
  *
  * The purpose of this macro is to help programmers find bugs in their
  * programs.
+ *
+ *
+ * With @ref DEBUG_ASSERT_VERBOSE defined this will print also the file, the
+ * line and the function this assertion failed in.
+ *
+ * @see http://pubs.opengroup.org/onlinepubs/9699919799/functions/assert.html
  */
-
-#ifdef NDEBUG
-#define assert(ignore)((void) 0)
+#define assert(cond) \
+    if (!(cond)) { \
+        printf("%s:%d: %s: Assertion '%s' failed\n", RIOT_FILE_RELATIVE, \
+                __LINE__, DEBUG_FUNC, #cond); \
+        core_panic(PANIC_ASSERT_FAIL, assert_crash_message); \
+    }
```

Comment:
Except for some cortex platforms (where we gain only about 20-52 byte) we seem to be able to scrape of a few 100 byte with this optimization, so I apply it